src/persist-client/src/cli/admin.rs

```diff
@@ -158,8 +181,102 @@ pub async fn run(command: AdminArgs) -> Result<(), anyhow::Error> {
+                    isolated_runtime,
+                    shared_states,
+                    pubsub_sender,
+                )?;
+                let diagnostics = Diagnostics {
+                    shard_name: shard_id.to_string(),
+                    handle_purpose: "persist-cli finalize shard".to_string(),
+                };
+
+                let mut write_handle: WriteHandle<
+                    crate::cli::inspect::K,
+                    crate::cli::inspect::V,
+                    u64,
+                    i64,
+                > = persist_client
+                    .open_writer(
+                        shard_id,
+                        Arc::new(TodoSchema::<crate::cli::inspect::K>::default()),
+                        Arc::new(TodoSchema::<crate::cli::inspect::V>::default()),
+                        diagnostics,
+                    )
+                    .await?;
+
+                if !write_handle.upper().is_empty() {
+                    let empty_batch: Vec<(
+                        (crate::cli::inspect::K, crate::cli::inspect::V),
+                        u64,
+                        i64,
+                    )> = vec![];
+                    let lower = write_handle.upper().clone();
+                    let upper = Antichain::new();
+
+                    let result = write_handle.append(empty_batch, lower, upper).await?;
+                    if let Err(err) = result {
+                        anyhow::bail!("failed to force downgrade upper, {err:?}");
+                    }
+                }
+            }
+
+            if force_downgrade_since {
+                let (state, _maintenance) = machine
+                    .register_critical_reader::<crate::cli::inspect::O>(
+                        &crate::PersistClient::CONTROLLER_CRITICAL_SINCE,
+                        "persist-cli finalize with force downgrade",
+                    )
+                    .await;
+
+                // HACK: Internally we have a check that the Opaque is using
+                // the correct codec. For the purposes of this command we want
+                // to side step that check so we set our reported codec to
+                // whatever the current state of the Shard is.
+                let expected_opaque = crate::cli::inspect::O::decode(state.opaque.0);
+                *FAKE_OPAQUE_CODEC.lock().expect("lockable") = state.opaque_codec.clone();
+
+                let (result, _maintenance) = machine
+                    .compare_and_downgrade_since(
+                        &crate::PersistClient::CONTROLLER_CRITICAL_SINCE,
+                        &expected_opaque,
+                        (&expected_opaque, &Antichain::new()),
+                    )
+                    .await;
+                if let Err((actual_opaque, _since)) = result {
+                    bail!(
+                        "opaque changed, expected: {expected_opaque:?}, actual: {actual_opaque:?}"
+                    )
+                }
+            }
```

Comment:
I'd be kind of interested why it was easier to use the high-level interfaces for the writer and the low-level ones for the reader, but it's definitely not a problem!